src/notebook/model.ts

```diff
@@ -642,209 +475,97 @@ class NotebookModel implements INotebookModel {
+      () => {
+        return this._metadata[name];
+      },
+      (value: string) => {
+        this.setCursorData(name, value);
+      }
     );
+    this._cursors.push(cursor);
+    return cursor;
   }
 
   /**
    * List the metadata namespace keys for the notebook.
-   *
-   * #### Notes
-   * Metadata associated with the nbformat are not included.
    */
   listMetadata(): string[] {
     return Object.keys(this._metadata);
   }
 
   /**
-   * Handle changes to cell editors.
-   */
-  protected onEditorChanged(editor: IEditorModel, args: IChangedArgs<any>): void {
-    if (args.name === 'text') {
-      this.dirty = true;
-    }
-  }
-
-  /**
-   * Handle an edge request from a text editor.
-   */
-  protected onEdgeRequested(sender: IEditorModel, args: EdgeLocation): void {
-    switch (args) {
-    case 'top':
-      this.activeCellIndex--;
-      break;
-    case 'bottom':
-      this.activeCellIndex++;
-      break;
-    }
-  }
-
-  /**
    * Handle a change in the cells list.
    */
-  protected onCellsChanged(list: ObservableList<ICellModel>, change: IListChangedArgs<ICellModel>): void {
+  protected onCellsChanged(list: IObservableList<ICellModel>, change: IListChangedArgs<ICellModel>): void {
     let cell: ICellModel;
     switch (change.type) {
     case ListChangeType.Add:
-      this.activeCellIndex = change.newIndex;
       cell = change.newValue as ICellModel;
-      cell.input.textEditor.edgeRequested.connect(this.onEdgeRequested, this);
-      cell.input.textEditor.stateChanged.connect(this.onEditorChanged, this);
+      cell.contentChanged.connect(this.onCellChanged, this);
       break;
     case ListChangeType.Remove:
-      (change.oldValue as ICellModel).dispose();
-      this.activeCellIndex = Math.min(change.oldIndex, this.cells.length - 1);
+      // Handled by undo.
       break;
     case ListChangeType.Replace:
-      let oldValues = change.oldValue as ICellModel[];
-      for (cell of oldValues) {
-        cell.dispose();
+      let newValues = change.newValue as ICellModel[];
+      for (cell of newValues) {
+        cell.contentChanged.connect(this.onCellChanged, this);
       }
       break;
+    case ListChangeType.Set:
+      cell = change.newValue as ICellModel;
+      cell.contentChanged.connect(this.onCellChanged, this);
+      break;
+    default:
+      return;
     }
     this.dirty = true;
   }
 
   /**
-   * The singleton callback for cursor change events.
+   * Handle a change to a cell state.
    */
-  private _cursorCallback(name: string): void {
-    this.metadataChanged.emit(name);
+  protected onCellChanged(cell: ICellModel, change: any): void {
+    this.contentChanged.emit('cells');
+  }
+
+  /**
+   * Set the cursor data for a given field.
+   */
+  protected setCursorData(name: string, value: string): void {
+    if (this._metadata[name] === value) {
+      return;
+    }
+    this._metadata[name] = value;
+    this.contentChanged.emit(`metadata.${name}`);
```

Comment:
Why is the contentChanged signal emitting a string (rather than a dict like IChangeArgs)? I realize that having a string avoids an object creation - is performance the concern? 

I think it would be cleaner if we didn't have to parse a string in order to get the name of the metadata attribute that changed. If we had an IChangedArgs or something, we could store the metadata changed field as a separate object attribute.